File: common/src/main/kotlin/net/impleri/playerskills/restrictions/PlayerConditions.kt
package net.impleri.playerskills.restrictions

import net.minecraft.resources.ResourceLocation
import java.util.function.Predicate
import net.minecraft.world.entity.player.Player as MCPlayer

interface PlayerConditions<Player> {
  val name: ResourceLocation?
  var target: String?
  var rawCondition: Predicate<Player>

  val actualCondition: (MCPlayer) -> Boolean

  fun condition(predicate: Predicate<Player>): PlayerConditions<Player> {
    rawCondition = Predicate { player: Player? -> player?.let { predicate.test(it) } ?: true }
    return this
  }

  fun unless(predicate: Predicate<Player>): PlayerConditions<Player> {
    rawCondition = Predicate { player: Player? -> player?.let { !predicate.test(it) } ?: true }
    return this
  }
}

File: common/src/main/scala/net/impleri/playerskills/restrictions/conditions/PlayerConditions.scala
package net.impleri.playerskills.restrictions.conditions

import net.impleri.slab.entity.Player

trait PlayerConditions {
  var condition: Player => Boolean = _ => true

  def predicate(predicate: Player => Boolean): Unit =
    condition = predicate

  def unless(predicate: Player => Boolean): Unit =
    condition = (player: Player) => !predicate(player)
}
